basic python api

charles-zhng opened the pull request:
Fixed PR for basic python api by creating new branch off of main and doing `git merge --squash [branch]`

[Issue 18](https://github.com/talmolab/stac-mjx/issues/18)

Minor refactoring to surface basic stac functionality in the form of two functions: 
- `load_config` loads the stac and model configs appropriately and returns the stac config
- `run_stac` takes the stac config and the user-prepared kp_data and runs the entirety of stac, with all options available via config changes. 

Testing for this can be seen in `use_api.ipynb` (should probably be renamed to something else). 

Also refactored CLI usage with the rodent example in `run_rodent.py` to live outside of main stac package.

TODOs before merge: testing for `load_config` and `run_stac`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a user-level API for running the STAC module.
	- Added a new function to load configuration settings from YAML files.

- **Configuration Updates**
	- Streamlined configuration paths for data processing.
	- Added new configuration files specific to rodent datasets.

- **Documentation**
	- Expanded README with clearer installation and usage instructions, including details on headless rendering setup.

- **Chores**
	- Removed unused entries in the `.gitignore` for better file tracking.
	- Updated test fixtures to improve testing reliability.
	- Created a new test file for validating configuration loading functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

charles-zhng commented on environment.yml:
```diff
@@ -17,9 +17,10 @@
   - glfw3
   - ipykernel
-  - pip: 
+  - pip:
```
I don't understand this one... 6 spaces would offset it from the rest of the list

jf514 commented on tests/test_main.py:
```diff
@@ -0,0 +1,46 @@
+from stac_mjx import main
+
+
+@pytest.fixture
+def mock_omegaconf_load(monkeypatch):
+    mock = Mock(
+        side_effect=[
+            OmegaConf.create({"model": "config"}),
+            OmegaConf.create({"stac": "config"}),
+        ]
+    )
+    monkeypatch.setattr(OmegaConf, "load", mock)
+    return mock
+
+
+@pytest.fixture
+def mock_init_params(monkeypatch):
+    mock = Mock()
+    monkeypatch.setattr(utils, "init_params", mock)
+    return mock
```
Do you think these will be useful for other tests, can you please put these in a file a `fixtures` dir. in tests.

charles-zhng replied:
I'm not sure if they'll be useful for other tests because testing things deeper in the logic will require actual param values, but will move to a fixtures dir.

jf514 reviewed (changes requested):
Overall looks good. Can you please describe the functionality in the README.md?

charles-zhng commented on stac_mjx/__init__.py:
```diff
@@ -1 +1 @@
-"""This module exposes all high level APIs for stac-mjx."""
+"""init file."""
```
No idea, will revert

charles-zhng commented on configs/rodent.yaml:
```diff
@@ -1,5 +1,14 @@
+N_FRAMES_PER_CLIP: 250
+
+# Tolerance for the optimizations of the full model, limb, and root. 
+# TODO: Re-implement optimizer loops to use these tolerances
```
https://github.com/talmolab/stac-mjx/issues/27

charles-zhng commented on tests/test_main.py:
```diff
@@ -0,0 +1,25 @@
+from omegaconf import OmegaConf, DictConfig
+from stac_mjx import main
+from tests.fixtures.configs import mock_omegaconf_load, mock_init_params
```
It is being used, bad coderabbit!

talmo commented on stac_mjx/controller.py:
```diff
@@ -305,3 +299,34 @@ def mjx_setup(kp_data, mj_model):
+
+    names_xpos = physics.named.data.xpos.axes.row.names
+
+    print(f"shape of qpos: {q.shape}")
```
remove debugging print statement or use `logging.info`

talmo commented on stac_mjx/main.py:
```diff
@@ -84,9 +86,9 @@ def run_stac(cfg: DictConfig):
+    if cfg.skip_transform == 1:
         logging.info("skipping transform()")
-        return fit_path, "none"
+        return fit_path, "No transform path"
```
```suggestion
        return fit_path, None
```

Checking for `NoneType` is a bit more idiomatic than string comparison.

talmo commented on use_api.ipynb:
Love this, but can we move these example notebooks into a `docs` directory?

charles-zhng replied:
Makes sense. In this case, `stac-mjx` would need to be appended to the path in order for the notebooks to run--is this alright? 
```
import sys
sys.path.append('../')
```

talmo replied:
If the package is installed (which it should be now that we have `pyproject.toml`), then it should be importable without having to add it to the path.

charles-zhng replied:
was able to pip install locally (`pip install .` in the directory), thanks! made the changes and pushed

talmo commented on videos/direct_render.mp4:
Should this really be committed to git history?

charles-zhng replied:
@jf514 Do we need this file? If not I can remove it

jf514 replied:
I think so. It's linked in the viz_usage notebook, so it gives the user an idea of correct output. I personally refer to it fairly often when confirming my changes haven't broken anything. Since it's useful to me, I imagine it would be useful to users.

charles-zhng replied:
okay, will keep for now then.

charles-zhng commented on docs/api_usage.ipynb:
```diff
@@ -0,0 +1,287 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/charles/miniconda3/envs/stac-mjx/lib/python3.12/site-packages/jaxopt/_src/optax_wrapper.py:120: DeprecationWarning: jax.tree_map is deprecated: use jax.tree.map (jax v0.4.25 or newer) or jax.tree_util.tree_map (any JAX version).\n",
+      "  return jax.tree_map(update_fun, params, updates)\n"
```
This is in Optax so ignoring

jf514 commented:
A couple of things came up during the merge: 

1. Tolerance is not used: [tolerance](https://github.com/talmolab/stac-mjx/blob/python-api-pr/stac_mjx/compute_stac.py#L54)
2. Create conda environment (w/ environment.yml) fails
3. test_main.py unit test should read in actual files

charles-zhng commented:
Latest commits include a few changes:
- merged in latest commits from main
- updated README and `api_usage.ipynb` to use new data loaders
- added more exception handling and corresponding unit tests for data loaders
- updated unit test for `load_configs()` to use provided test config files

jf514 reviewed (approved):
LGTM. Thanks for your hard work!
